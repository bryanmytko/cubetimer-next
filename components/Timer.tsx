import React, { useEffect, useReducer } from "react";

import Scrambler from "../lib/scrambler";
import { TimerReducer, TimerActionKind } from "../reducers";
/* import { FormatService, ScrambleService } from '../../services'; */

//const { humanReadableTime } = FormatService;

interface TimerState {
  running: boolean;
  ready: boolean;
  time: number;
  solveTimes: Array<number>;
  scramble: string;
}

const initialState: TimerState = {
  running: false,
  ready: false,
  time: 0,
  solveTimes: [],
  scramble: new Scrambler("3x3").generate(),
};

const Timer = () => {
  const [state, dispatch] = useReducer(TimerReducer, initialState);

  useEffect(() => {
    let interval: NodeJS.Timeout;

    if (state.running) {
      interval = setInterval(
        () => dispatch({ type: TimerActionKind.TICK }),
        10
      );
    }

    const handleKeydown = (e: KeyboardEvent) => {
      if (e.key !== " ") return;
      e.preventDefault();
      dispatch({ type: TimerActionKind.READY });
    };

    const handleKeyup = (e: KeyboardEvent) => {
      if (e.key !== " ") return;
      e.preventDefault();
      dispatch({ type: TimerActionKind.TOGGLE });
    };

    window.addEventListener("keydown", handleKeydown);
    window.addEventListener("keyup", handleKeyup);

    return () => {
      clearInterval(interval);
      window.removeEventListener("keydown", handleKeydown);
      window.removeEventListener("keyup", handleKeyup);
    };
  }, [state.running]);

  return (
    <div>
      <div className="card rounded bg-gray-50 dark:bg-slate-700 py-4 mx-auto mt-6 text-center w-4/5">
        <span className="text-6xl text-white">{state.time}</span>
      </div>
      <button
        id="timer-btn"
        className={`timer-btn-start block mx-auto mt-6 px-20 py-5 rounded-md ${state.ready ? "dark:bg-red-500" : "dark:bg-green-500"
          }`}
        onClick={() => dispatch({ type: TimerActionKind.TOGGLE })}
      >
        Press spacebar or click to begin!
      </button>
    </div>
  );
};

export default Timer;
