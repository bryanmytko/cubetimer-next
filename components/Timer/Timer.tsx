import { useEffect, useReducer } from "react";

import { Scramble } from "../";
import Panel from "./Panel";
import Times from "./Times";
import { humanReadableTime } from "../../lib/format";
import { TimerReducer, TimerActionKind } from "../../reducers";
import { TimerState } from "../../types/timer";

const Timer = () => {
  const initialState: TimerState = {
    countdown: 0,
    inspectionRunning: false,
    running: false,
    ready: false,
    time: 0,
    solveTimes: [],
    scramble: "",
    inspectionTime: 0,
  };

  const [state, dispatch] = useReducer(TimerReducer, initialState);

  const handleKeyup = (e: KeyboardEvent) => {
    if (e.key !== " ") return;
    e.preventDefault();
    dispatch({ type: TimerActionKind.TOGGLE_RUNNING });
  };

  const handleKeydown = (e: KeyboardEvent) => {
    if (e.key !== " ") return;
    e.preventDefault();
    dispatch({ type: TimerActionKind.READY });
  };

  /* This is necessary due to the random nature of the initial scramble.
   * Just trying to set an initial state with a random scramble will
   * cause hydration errors on rerender due to mismatched text
   */
  useEffect(() => {
    dispatch({ type: TimerActionKind.INITIALIZE });
  }, []);

  useEffect(() => {
    let interval: NodeJS.Timeout;

    const inspectionTimer = () => {
      dispatch({
        type: TimerActionKind.COUNTDOWN,
        value: state.countdown--,
      });

      state.countdown >= 0
        ? setTimeout(inspectionTimer, 1000)
        : dispatch({ type: TimerActionKind.TOGGLE_INSPECTION });
    };

    const timer = () => {
      dispatch({
        type: TimerActionKind.COUNTDOWN,
        value: state.inspectionTime,
      });

      interval = setInterval(
        () => dispatch({ type: TimerActionKind.TICK_UP }),
        60
      );
    };

    if (state.running) {
      if (!state.inspectionRunning && state.countdown > 0) {
        dispatch({ type: TimerActionKind.TOGGLE_INSPECTION });
      } else if (state.inspectionRunning && state.countdown > 0) {
        inspectionTimer();
      } else {
        timer();
      }
    }

    window.addEventListener("keydown", handleKeydown);
    window.addEventListener("keyup", handleKeyup);

    return () => {
      clearInterval(interval);
      window.removeEventListener("keydown", handleKeydown);
      window.removeEventListener("keyup", handleKeyup);
    };
  }, [state.running, state.inspectionRunning]);

  return (
    <>
      <div className="col-span-5">
        <div>
          <Scramble scramble={state.scramble} />
          <div
            className={`w-11/12 py-4 mx-auto mt-6 text-center rounded card ${
              state.inspectionRunning ? "bg-red-500" : "bg-slate-700"
            }`}
          >
            <span
              id="timer-screen"
              className="font-mono font-black text-white timer text-8xl"
            >
              {state.running && state.inspectionRunning
                ? state.countdown
                : humanReadableTime(state.time, "0:00")}
            </span>
          </div>
          <button
            id="timer-btn"
            className={`timer-btn-start block mx-auto mt-6 px-10 py-5 text-3xl rounded-md w-11/12 ${
              state.ready ? "bg-red-500" : "bg-yellow-300"
            }`}
            onClick={() => dispatch({ type: TimerActionKind.TOGGLE_RUNNING })}
          >
            Press spacebar or click to start!
          </button>
          <Panel dispatch={dispatch} solveTimes={state.solveTimes} />
        </div>
      </div>
      <Times dispatch={dispatch} solveTimes={state.solveTimes} />
    </>
  );
};

export default Timer;
