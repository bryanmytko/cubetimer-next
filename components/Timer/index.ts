export { default as Clock } from "./Clock";
export { default as ClockButton } from "./ClockButton";
export { default as Panel } from "./Panel";
export { default as Timer } from "./Timer";
export { default as Times } from "./Times";
