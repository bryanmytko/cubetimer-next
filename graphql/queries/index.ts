export { SOLVES_FOR_USER } from "./solvesForUser";