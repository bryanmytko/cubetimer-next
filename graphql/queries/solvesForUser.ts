import { gql } from "@apollo/client";

export const SOLVES_FOR_USER = gql`
  query {
    solves {
      id
      puzzle
      scramble
      time
    }
  }
`;