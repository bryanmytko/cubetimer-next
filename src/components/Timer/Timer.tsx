import { TimerProvider } from "./TimerContext";
import { TimerContainer } from "./";

const Timer = () => {
  return (
    <TimerProvider>
      <TimerContainer />
    </TimerProvider>
  );
};

export default Timer;
