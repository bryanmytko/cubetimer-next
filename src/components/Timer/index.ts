export { default as ClassicModeTimes } from "./ClassicModeTimes";
export { default as Clock } from "./Clock";
export { default as ClockButton } from "./ClockButton";
export { default as ConfirmModal } from "./ConfirmModal";
export { default as HumanReadableTime } from "./HumanReadableTime";
export { default as Panel } from "./Panel";
export { default as Timer } from "./Timer";
export { default as TimerContainer } from "./TimerContainer";
export { default as Times } from "./Times";
