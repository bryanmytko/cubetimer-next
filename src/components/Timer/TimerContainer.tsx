import { Dispatch, useCallback, useContext, useEffect, useState } from "react";
import useSound from "use-sound";
import { useSession } from "next-auth/react";
import { useMutation, useQuery } from "@apollo/client";
import dynamic from "next/dynamic";

const DynamicTimes = dynamic(() => import("./Times"));
const DynamicClassicModeTimes = dynamic(() => import("./ClassicModeTimes"));
const DynamicClock = dynamic(() => import("./Clock"));
const DynamicClockButton = dynamic(() => import("./ClockButton"));
const DynamicConfirmModal = dynamic(() => import("./ConfirmModal"));
const DynamicPanel = dynamic(() => import("./Panel"));
const DynamicScramble = dynamic(() => import("../Scramble/Scramble"));

import { CREATE_SOLVE_SESSION, SAVE_SOLVE } from "../../graphql/mutations";
import {
  RecordSolveOptions,
  TimerAction,
  TimerActionKind,
  TimerState,
} from "../../types/timer";
import { TimerContext, TimerDispatchContext } from "./TimerContext";
import { useTimerKeyup } from "../../hooks";
import { SETTINGS_FOR_USER } from "../../graphql/queries/settingsForUser";

const AUDIO_DING = "/assets/audio/ding.mp3";
const TICK = 60;

const TimerContainer = () => {
  const [saveSolve, { loading: saveTimeLoading }] = useMutation(SAVE_SOLVE);
  const [createSolveSession, { loading: createSessionLoading }] =
    useMutation(CREATE_SOLVE_SESSION);
  const { data: session, status } = useSession();
  const [playDing] = useSound(AUDIO_DING);
  const timer = useContext(TimerContext) as TimerState;
  const dispatch = useContext(TimerDispatchContext) as Dispatch<TimerAction>;
  const [submitted, setSubmitted] = useState(false);
  const { data, loading } = useQuery(SETTINGS_FOR_USER, {
    skip: !session?.user?.id,
    variables: { userId: session?.user?.id },
  });

  /* This is necessary due to the random nature of the initial scramble.
   * Just trying to set an initial state with a random scramble will
   * cause hydration errors on rerender due to mismatched text
   */
  useEffect(() => {
    dispatch({
      type: TimerActionKind.INITIALIZE,
    });
  }, [dispatch]);

  useEffect(() => {
    if (data?.settingsForUser?.defaultClassicMode)
      dispatch({ type: TimerActionKind.TOGGLE_CLASSIC_MODE });
  }, [data, dispatch]);

  const setButtonLocked = useCallback(
    (value: boolean) => dispatch({ type: TimerActionKind.LOCK, value }),
    [dispatch],
  );

  useEffect(() => {
    timer.classicModeEnabled &&
    timer.solveTimes.length >= timer.classicModeLength
      ? setButtonLocked(true)
      : setButtonLocked(false);
  }, [
    setButtonLocked,
    timer.solveTimes,
    timer.classicModeEnabled,
    timer.classicModeLength,
  ]);

  const toggleConfirmModal = useCallback(() => {
    dispatch({ type: TimerActionKind.TOGGLE_CONFIRM_ACTIVE });
    setButtonLocked(!timer.locked);
  }, [dispatch, setButtonLocked, timer.locked]);

  const getSolveSession = useCallback(
    async ({ userId }: { userId: string }) => {
      if (timer.classicModeEnabled && !timer.solveSessionId) {
        const response = await createSolveSession({
          variables: { userId, size: timer.classicModeLength },
        });
        dispatch({
          type: TimerActionKind.SET_SOLVE_SESSION_ID,
          id: response.data?.createSolveSession.id,
        });
        return response.data?.createSolveSession.id;
      }

      return timer.solveSessionId;
    },
    [
      createSolveSession,
      dispatch,
      timer.classicModeEnabled,
      timer.solveSessionId,
      timer.classicModeLength,
    ],
  );

  const recordSolve = useCallback(
    async (options: RecordSolveOptions = { penalty: 0 }) => {
      const { penalty } = options;
      setButtonLocked(true);
      let solveId;

      if (session) {
        const userId = session.user.id;
        const solveSessionId = await getSolveSession({ userId });
        const response = await saveSolve({
          variables: {
            penalty,
            puzzle: timer.puzzleType,
            scramble: timer.scramble,
            time: timer.time,
            userId,
            solveSessionId,
          },
        });

        solveId = response.data?.createSolve.id;
      }

      dispatch({
        type: TimerActionKind.ADD_TIME,
        penalty,
        solveId,
        scramble: timer.scramble,
      });
      toggleConfirmModal();
      setSubmitted(false);
      setButtonLocked(false);
    },
    [
      dispatch,
      getSolveSession,
      saveSolve,
      session,
      setButtonLocked,
      timer.puzzleType,
      timer.scramble,
      timer.time,
      toggleConfirmModal,
    ],
  );

  const handleKeydown = useCallback(
    (e: KeyboardEvent) => {
      if (e.key === " ") e.preventDefault();
      if (e.key !== " " || timer.locked || submitted) return;

      dispatch({ type: TimerActionKind.READY });
    },
    [dispatch, submitted, timer.locked],
  );

  const handleEscape = useCallback(() => {
    if (timer.confirmActive) {
      dispatch({ type: TimerActionKind.CANCEL_SOLVE });
    }
  }, [dispatch, timer.confirmActive]);

  const handleEnter = useCallback(() => {
    if (submitted) return;
    if (timer.confirmActive) {
      setSubmitted(true);
      recordSolve();
    }
  }, [timer.confirmActive, recordSolve, submitted]);

  const handleSpacebar = useCallback(() => {
    if (timer.locked) return handleEnter();

    timer.inspectionTime && !timer.running
      ? dispatch({ type: TimerActionKind.TOGGLE_INSPECTION })
      : dispatch({ type: TimerActionKind.TOGGLE_RUNNING });

    if (!timer.running) return;

    toggleConfirmModal();
  }, [
    dispatch,
    timer.locked,
    handleEnter,
    timer.inspectionTime,
    timer.running,
    toggleConfirmModal,
  ]);

  const timerKeyup = useTimerKeyup(timer.keyLocked, dispatch, {
    " ": handleSpacebar,
    Escape: handleEscape,
    Enter: handleEnter,
  });

  useEffect(() => {
    window.addEventListener("keydown", handleKeydown);
    window.addEventListener("keyup", timerKeyup);

    return () => {
      window.removeEventListener("keydown", handleKeydown);
      window.removeEventListener("keyup", timerKeyup);
    };
  }, [handleKeydown, timerKeyup]);

  useEffect(() => {
    let interval: NodeJS.Timeout;

    if (timer.inspectionRunning) {
      const countDown = () => {
        if (timer.countdown > 1) {
          dispatch({
            type: TimerActionKind.COUNTDOWN,
            value: --timer.countdown,
          });
        } else {
          clearInterval(interval);
          playDing();
          dispatch({ type: TimerActionKind.TOGGLE_RUNNING });
        }
      };

      interval = setInterval(() => countDown(), 1000);
      return;
    }

    dispatch({
      type: TimerActionKind.COUNTDOWN,
      value: timer.inspectionTime,
    });

    return () => clearInterval(interval);
  }, [
    dispatch,
    setButtonLocked,
    timer.countdown,
    timer.inspectionRunning,
    timer.inspectionTime,
    playDing,
  ]);

  useEffect(() => {
    let interval: NodeJS.Timeout;

    if (timer.running) {
      interval = setInterval(
        () => dispatch({ type: TimerActionKind.TICK_UP }),
        TICK,
      );
    }

    return () => clearInterval(interval);
  }, [dispatch, timer.running]);

  if (loading || status === "loading") return <></>;

  return (
    <>
      <div className={timer.classicModeEnabled ? "col-span-6" : "col-span-5"}>
        <DynamicScramble />
        <DynamicClock />
        <DynamicClockButton handleKeyup={timerKeyup} />
        {timer.classicModeEnabled && <DynamicClassicModeTimes />}
        <DynamicConfirmModal
          action={recordSolve}
          loading={saveTimeLoading || createSessionLoading}
        />
        <DynamicPanel initialClassicModeEnabled={timer.classicModeEnabled} />
      </div>
      {!timer.classicModeEnabled && <DynamicTimes session={session} />}
    </>
  );
};

export default TimerContainer;
