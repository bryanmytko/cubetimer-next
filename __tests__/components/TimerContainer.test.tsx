import { render, waitFor } from "@testing-library/react";
import { MockedProvider } from "@apollo/client/testing";
import "@testing-library/jest-dom";

import Scrambler, { mockGenerate } from "../__mocks__/scrambler";
import TimerContainer from "../../src/components/Timer/TimerContainer";
import { TimerProvider } from "../../src/components/Timer/TimerContext";
import { SOLVES_FOR_USER } from "../../src/graphql/queries/solvesForUser";

jest.mock("../../src/lib/scrambler", () => Scrambler);

jest.mock("next-auth/react", () => {
  const originalModule = jest.requireActual("next-auth/react");
  const mockSession = {
    expires: new Date(Date.now() + 2 * 86400).toISOString(),
    user: { username: "admin", id: "12345687890" },
  };
  return {
    __esModule: true,
    ...originalModule,
    useSession: jest.fn(() => {
      return { data: mockSession, status: "authenticated" };
    }),
  };
});

const mocks = [
  {
    request: {
      query: SOLVES_FOR_USER,
      variables: { userId: "12345687890", first: 20 },
    },
    result: {
      data: {
        solves: {
          pageInfo: {
            startCursor: "R1BDOk46MTczNzc=",
            endCursor: "R1BDOk46MTczNTg=",
            hasPreviousPage: false,
            hasNextPage: true,
            __typename: "PageInfo",
          },
          edges: [
            {
              cursor: "R1BDOk46MTczNzc=",
              node: {
                id: "17377",
                puzzle: "3x3",
                scramble:
                  "F D2 F2 U' B D' R F D2 R2 F' L B L' D' L' U R' U2 L B",
                time: 14760,
                penalty: 0,
                createdAt: "2025-01-27",
                __typename: "Solve",
              },
              __typename: "QuerySolvesConnectionEdge",
            },
            {
              cursor: "R1BDOk46MTczNzY=",
              node: {
                id: "17376",
                puzzle: "3x3",
                scramble:
                  "R B' R' B L' B' D B2 R U2 B U L D2 B2 L2 F L2 B' U2 F'",
                time: 20040,
                penalty: 0,
                createdAt: "2025-01-27",
                __typename: "Solve",
              },
              __typename: "QuerySolvesConnectionEdge",
            },
          ],
          __typename: "QuerySolvesConnection",
        },
      },
    },
  },
];

describe("<TimerContainer>", () => {
  beforeEach(() => {
    Scrambler.mockClear();
    mockGenerate.mockClear();
  });

  it("renders the components", async () => {    
    const { container } = render(
      <MockedProvider mocks={mocks}>
        <TimerProvider>
          <TimerContainer />
        </TimerProvider>
      </MockedProvider>
    );

    await waitFor(() => expect(container).toMatchSnapshot());
  });
});
