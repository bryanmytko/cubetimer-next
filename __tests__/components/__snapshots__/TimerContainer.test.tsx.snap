// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TimerContainer> renders the components 1`] = `
<div>
  <div
    class="col-span-6"
  >
    <div
      class="flex items-center justify-center gap-2 w-full h-dvh mt-[-220px]"
    >
      <div
        class="w-20 h-20"
      >
        <svg
          viewBox="0 0 200 200"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            fill="#135F75"
            height="30"
            stroke="#135F75"
            stroke-width="15"
            width="30"
            x="25"
            y="85"
          >
            <animate
              attributeName="opacity"
              begin="-.4"
              calcMode="spline"
              dur="1.3"
              keySplines=".5 0 .5 1;.5 0 .5 1"
              repeatCount="indefinite"
              values="1;0;1;"
            />
          </rect>
          <rect
            fill="#135F75"
            height="30"
            stroke="#135F75"
            stroke-width="15"
            width="30"
            x="85"
            y="85"
          >
            <animate
              attributeName="opacity"
              begin="-.2"
              calcMode="spline"
              dur="1.3"
              keySplines=".5 0 .5 1;.5 0 .5 1"
              repeatCount="indefinite"
              values="1;0;1;"
            />
          </rect>
          <rect
            fill="#135F75"
            height="30"
            stroke="#135F75"
            stroke-width="15"
            width="30"
            x="145"
            y="85"
          >
            <animate
              attributeName="opacity"
              begin="0"
              calcMode="spline"
              dur="1.3"
              keySplines=".5 0 .5 1;.5 0 .5 1"
              repeatCount="indefinite"
              values="1;0;1;"
            />
          </rect>
        </svg>
      </div>
    </div>
  </div>
</div>
`;
